### ISSUE TITLE
Inaccurate location information for VariableDeclaration since 0.13.x

### ISSUE THREAD

**Issue by tintinweb** [solidity-parser/parser#46]
Hey 👋 

I am using this parser in my vscode extension [vscode-visual-developer](https://marketplace.visualstudio.com/items?itemName=tintinweb.solidity-visual-auditor). I just tried updating von 0.12.x to 0.13.x and found out that the location information for `VariableDeclaration` AST types is off. It seems to always start at line-start now 🤔 

Here's a screenshot that visualizes this. My extension highlights state-vars by drawing a golden dotted box around state-vars (and their declaration). Since 0.13.0 it highlights the full line (starting from col:1 to end of the declaration). 

This may be a bug in the parser. Would be great if you could investigate.

cheers,
tin

Actual: (complete declaration line highlighted: `VariableDeclaration.loc` seems to be off.

![image](https://user-images.githubusercontent.com/2865694/119888923-3dbf0480-bf36-11eb-9c1c-e97c5aeddcb9.png)

Expected: behavior from 0.12.x. `VariableDeclaration.loc` returns range for the variable name only.

**Comment by fvictorio** [comment]
Ugh, I did a huge refactor for 0.13 and I might have fucked up something. Thanks for reporting this!

**Comment by fvictorio** [comment]
Hmm, I'm looking into this and I feel that the _previous_ behavior was the buggy one. Since the node is a `VariableDeclaration`, then the whole declaration should be the range of the node.

That being said, I understand why you'd want to know the range of the name of the declared variable. That's not possible right now because the node has a `name: string` property, instead of having a child that is an identifier. Something similar happens with function calls with the `{ }` syntax.

The tl;dr is: would you be able to do what you want to do if the node was like this:

```
{
  type: "VariableDeclaration",
  name: "foo",
  identifier: {
    type: "Identifier",
    name: "foo",
    loc: { ... }
  }
}
```

This new `identifier` property is somewhat redundant with the `name` property, but that's fine (that's the way I've been keeping things backwards-compatible so far).

**Comment by tintinweb** [comment]
I think I get your point. `VariableDeclaration` refers to the complete line and hence `loc` should reflect that. There's already some redundancy with the `StateVariableDeclaration.variables=[VariableDeclaration, ...]` pattern (unless one svar declaration can really reference multiple vardecs). 

The identifier as proposed would be an option that would definitely work for me and would give direct access to location information for the vardec-identifier 🙌.

//for ref here's the structure of a `StateVariableDeclaration` `->` `VariableDeclaration`
![image](https://user-images.githubusercontent.com/2865694/120011901-1a9c5f80-bfdf-11eb-8106-babf23468253.png)

**Comment by fvictorio** [comment]
Yes, I don't know why `StateVariableDeclaration` is always a singleton array with `VariableDeclaration`, I think it was done that way before I started maintaining the parser. At some point I will make a huge breaking change revamping the AST and hopefully the structure will be a little better.

**Comment by fvictorio** [comment]
Hey @tintinweb, I just published a new version that includes this change. Please let me know if it works for you and feel free to re-open this if it doesn't.

Btw, I want to create a Telegram for users of this parser. Do you mind if I add you? (I assume your telegram handle is the one in your github profile).

**Comment by tintinweb** [comment]
@fvictorio cool 😎 count me in. is this the solc-tooling telegram channel or a solidity-parser specific one?

btw. I am running into an infinite recursion loop with v0.13.0 while parsing `Identifiers` from a `functionDefinition` node. I traced it down to this line: 

https://github.com/solidity-parser/parser/blame/1431a8eeb6b863aa3b8521214d2896141d7509ec/src/parser.ts#L121

It checks for `false` but in my case this may return `undefined` (I must confess I am enriching astNodes with some extra references to other nodes 😬 ). It seems to be working if I check for `undefined` or a simple `false` check and it was working with v0.12.x.

here's a screenshot from my lib:
![image](https://user-images.githubusercontent.com/2865694/120081531-696cf680-c0be-11eb-9a90-73a998048539.png)

**Comment by fvictorio** [comment]
That is very weird. Do you think you can make a minimal reproduction example of this?

If not, would just checking for undefined there (besides false) solve your problem?

**Comment by tintinweb** [comment]
@fvictorio ignore that last issue. it magically resolved. probably a vscode issue or a dirty state. I'll open an issue if it happens again and will provide a minimal testcase.

cheers 🙌

**Comment by tintinweb** [comment]
digged a little deeper into how the visitor works to debug the initial problem I had. this is what I found out:

* The check at https://github.com/solidity-parser/parser/blame/1431a8eeb6b863aa3b8521214d2896141d7509ec/src/parser.ts#L121 essentially allows a type handler function to signal that parsing should stop at this layer. If we would have changed this to `undefined` we'd have broken the parser. It would basically stop traversing the tree too early.
* I wrapped my "extra information" that I am adding to astnodes while traversing the tree in an object. That basically stops the visitor from traversing into any astNodes I had referenced there. (e.g. some back-references from identifiers to their declarations aO).

Important learning: The visitor checks if the node is an Array and recurses into all elements. If the node is not an array  and not an object that contains a type field it stops. Also, you can set callbacks for any property name which may be handy. On the flip side that's not really efficient. (I initially thought it was a bit smarter :D)

just leaving this here in case someone else tapping into this.
